### ISSUE TITLE
istioctl pc listeners(clusters) add output whether its inbound or outbound

### ISSUE THREAD

**Issue by incfly** [istio/istio#19394]
**Describe the feature request**

Currently all listeners and clusters are listed as the same, without distinguishing as inboud or outbound.

Having it check the name prefix, then output the type can be helpful.

```
istioctl pc listeners $(kpidn bug -lapp=httpbin1).bug

ADDRESS           PORT      TYPE
10.40.5.152       15020     TCP
10.43.240.153     9000      TCP
10.43.251.212     443       TCP
10.43.240.10      53        TCP
10.43.247.222     443       TCP
10.43.244.208     443       TCP
10.43.251.212     15443     TCP
10.43.248.29      443       TCP
10.43.252.124     15011     TCP
10.43.240.1       443       TCP
10.43.247.222     15443     TCP
0.0.0.0           9091      TCP
0.0.0.0           15004     TCP
10.43.246.123     14268     TCP
10.43.243.106     16686     TCP
0.0.0.0           15019     TCP
10.43.247.222     15020     TCP
0.0.0.0           15010     TCP
10.43.255.100     42422     TCP
10.43.245.213     80        TCP
10.43.246.123     14250     TCP
10.43.246.123     14267     TCP
0.0.0.0           20001     TCP
0.0.0.0           15014     TCP
10.43.247.222     15032     TCP
0.0.0.0           9090      TCP
0.0.0.0           3000      TCP
0.0.0.0           80        TCP
10.43.247.222     15031     TCP
10.43.250.249     443       TCP
0.0.0.0           8060      TCP
10.43.247.222     15030     TCP
10.43.247.222     15029     TCP
0.0.0.0           9411      TCP
0.0.0.0           8080      TCP
10.43.240.28      443       TCP
0.0.0.0           8000      TCP
0.0.0.0           9901      TCP
0.0.0.0           15001     TCP
0.0.0.0           15006     TCP
0.0.0.0           15090     HTTP
```

**Additional context**

**Comment by howardjohn** [comment]
We will need bigger changes soon anyways once we move to filter chains cc @lambdai

**Comment by incfly** [comment]
Got it. Then I would re-word this as distinguish inbound and outbound config for `istioctl pc`, whether it's filter chain or listeners, making this difference always help IMO.

**Comment by lambdai** [comment]
IMHO inbound/outbound and my filter chain change is orthogonal
The reason is that each listener/filter has the attribute: inbound or outbound. It is never early to add the in/out attribute.

**Comment by lambdai** [comment]
WRT to the fact of listener when outbound listener only have 1 physical listener :15001
I think we might want to maintain the current listener *view* and hide the detail of the only port 15001 listener 
WDTY @howardjohn

**Comment by incfly** [comment]
It depends on how the intended UX for our istioctl pc debugging tool should be. Whether it's for seasoned users to drill lower level Envoy config or we still need to maintain another higher level view.

I think we can have both. Displaying filter chain && listener, plus how are they  selected by the filter chain match attributes, in a priority order, destination port, ip address, etc.

@esnible might be able to provide some comments.

**Comment by esnible** [comment]
We want to have commands that can be used for power users to quickly illustrate something that has gone wrong, and also tools to help new users understand what is happening.

Is there anything we can do about those addresses?  To help users understand what they are, and why so many addresses are being listened upon?

**Comment by incfly** [comment]
0.0.0.0_port is for outbound listener, for http
<cluster-ip>:svc-port is for tcp, outbound, listener.
<pod-ip>:container-port is for inbound listener
virtualInbound is for all other default to TCP proxy case.

Having said that, I think we need a more reliable way to distinguish inbound/outbound traffic. Above is just networking impl details, and subject to change. It turns out not a good practice to rely on. not sure if we can have proto level labeling/annotation on listener debugging annotation for this kinda of information.

**Comment by incfly** [comment]
i admit this is a thorny thing, in terms of making config understandable, while we have upcoming envoy config change.

are there some ways to generate a request handling decision tree, maybe? i think that's a good analogy based on my understanding of the listener/filter chain matching logic. the only thing is how to make generated config keep up with the istioctl tooling. as mentioned above, re-interpret again in istioctl is not something we should be doing.

**Comment by howardjohn** [comment]
on the listener there is a field "direction" which we set and can use

**Comment by esnible** [comment]
Would it bind Istioctl too tightly to Kubernetes if the `pc listeners` had a list of IPs (derived by using client-go equivalent of `kubectl get svc -o custom-columns=NAME:.metadata.name,CLUSTER-IP:.spec.clusterIP` and `kubectl get pods -o wide`?

It would be awesome if the output was something like
```
ADDRESS           WHAT          PORT      TYPE
10.40.5.152       pod:foo         15020     TCP
10.43.240.153     service:bar         9000      TCP
10.43.251.212     pod:baz         443       TCP
10.43.240.10      pod:bif         53        TCP
0.0.0.0           outbound     15006     TCP
0.0.0.0           outbound     15090     HTTP
```

**Comment by howardjohn** [comment]
@esnible if we want something like that I suggest we just have the WHAT say the cluster name we are sending to. There are a lot of complications with what you described, for external services, subsets, etc. And we aren't tied to k8s, and you can use `istioctl pc clusters` to map it

**Comment by esnible** [comment]
Cluster name should be sufficient, agreed.

**Comment by esnible** [comment]
@incfly I am considering giving this the `community/intern-help-wanted` tag.  Can you summarize what you now believe this subcommand's output should look like for 1.7?

**Comment by howardjohn** [comment]
This is implemented by `--verbose` on listener, and cluster now shows direction
